File: 3dwebshop-backend/3dwebshop-backend/src/main/java/com/my3dwebshop/_dwebshop_backend/service/ReviewService.java
package com.my3dwebshop._dwebshop_backend.service;

import com.my3dwebshop._dwebshop_backend.model.Product;
import com.my3dwebshop._dwebshop_backend.model.Review;
import com.my3dwebshop._dwebshop_backend.model.User;
import com.my3dwebshop._dwebshop_backend.repository.ReviewRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

/**
 * Service class for review management operations.
 * Handles creating, retrieving, and deleting product reviews.
 */
@Service
@RequiredArgsConstructor
public class ReviewService {

    private final ReviewRepository reviewRepository;

    /**
     * Retrieves all reviews.
     * 
     * @return List of all reviews
     */
    public List<Review> getAll() {
        return reviewRepository.findAll();
    }

    /**
     * Creates a new review.
     * 
     * @param review Review entity to create
     * @return Created review
     */
    public Review create(Review review) {
        return reviewRepository.save(review);
    }

    /**
     * Retrieves all reviews for a specific product.
     * 
     * @param product Product entity
     * @return List of reviews for the product
     */
    public List<Review> getByProduct(Product product) {
        return reviewRepository.findByProduct(product);
    }

    /**
     * Retrieves all reviews by a specific user.
     * 
     * @param user User entity
     * @return List of reviews by the user
     */
    public List<Review> getByUser(User user) {
        return reviewRepository.findByUser(user);
    }

    /**
     * Finds a review by user and product.
     * 
     * @param user User entity
     * @param product Product entity
     * @return Optional containing review if found
     */
    public Optional<Review> getByUserAndProduct(User user, Product product) {
        return reviewRepository.findByUserAndProduct(user, product);
    }

    /**
     * Deletes a review by ID.
     * 
     * @param id Review ID to delete
     */
    public void delete(Long id) {
        reviewRepository.deleteById(id);
    }
}

File: 3dwebshop-frontend/src/services/ReviewService.js
import api from "../axios";

/**
 * Review service for managing product reviews.
 * Handles creating, retrieving, and deleting reviews.
 */
const ReviewService = {
    /**
     * Fetches all reviews (for homepage).
     * 
     * @returns {Promise<Array>} Promise resolving to array of reviews
     * @throws {Error} If request fails
     */
    getAllReviews: async () => {
        try {
            const response = await api.get("/api/reviews");
            return response.data;
        } catch (error) {
            console.error("Error fetching all reviews:", error);
            throw error;
        }
    },

    /**
     * Fetches reviews for a specific product.
     * 
     * @param {number|string} productId - Product ID
     * @returns {Promise<Array>} Promise resolving to array of reviews
     * @throws {Error} If request fails
     */
    getByProduct: async (productId) => {
        try {
            const response = await api.get(`/api/reviews/product/${productId}`);
            return response.data;
        } catch (error) {
            console.error("Error fetching product reviews:", error);
            throw error;
        }
    },

    /**
     * Fetches reviews by a specific user.
     * 
     * @param {number|string} userId - User ID
     * @returns {Promise<Array>} Promise resolving to array of reviews
     * @throws {Error} If request fails
     */
    getByUser: async (userId) => {
        try {
            const response = await api.get(`/api/reviews/user/${userId}`);
            return response.data;
        } catch (error) {
            console.error("Error fetching user reviews:", error);
            throw error;
        }
    },

    /**
     * Creates a new review for a product.
     * 
     * @param {number|string} userId - User ID creating the review
     * @param {number|string} productId - Product ID being reviewed
     * @param {Object} reviewData - Review data (rating, comment)
     * @returns {Promise<Object>} Promise resolving to created review
     * @throws {Error} If request fails
     */
    createReview: async (userId, productId, reviewData) => {
        try {
            const response = await api.post(`/api/reviews/${userId}/${productId}`, reviewData);
            return response.data;
        } catch (error) {
            console.error("Error creating review:", error);
            throw error;
        }
    },

    /**
     * Deletes a review.
     * 
     * @param {number|string} reviewId - Review ID
     * @returns {Promise} Axios response
     * @throws {Error} If request fails
     */
    deleteReview: async (reviewId) => {
        try {
            await api.delete(`/api/reviews/${reviewId}`);
        } catch (error) {
            console.error("Error deleting review:", error);
            throw error;
        }
    }
};

export default ReviewService;

